compiler/rustc_trait_selection/src/traits/const_evaluatable.rs

```diff
@@ -185,14 +185,20 @@ pub fn is_const_evaluatable<'cx, 'tcx>(
         }
         let concrete = infcx.const_eval_resolve(param_env, uv.expand(), Some(span));
         match concrete {
-            Err(ErrorHandled::TooGeneric) => Err(if !uv.has_infer_types_or_consts() {
+            Err(ErrorHandled::TooGeneric) => Err(if uv.has_infer_types_or_consts() {
```

Comment:
`TooGeneric` here is a bit of a misnomer, since it really should be "not concrete" or something.

Reply:
yeah, so `InferCtxt::const_eval_resolve` is not not doing what we need for this
https://github.com/rust-lang/rust/blob/03d488b48af9f66b91e9400387f781b82411fa82/compiler/rustc_infer/src/infer/mod.rs#L1664-L1671

I guess what we should do is:
match on `AbstractConst::new(self.tcx, unevaluated.shrink());`

`Ok(None)` is already correct.
`Err(emitted) ` -> `NotConstEvaluatable::Error(emitted)`
`Ok(Some(x))` -> the inference vars aren't used in the anon const, replace them with placeholders

feel free to do that in this PR, otherwise r=me

Reply:
@lcnr: I'm gonna do it in a follow-up PR, since this solution doesn't really _rely_ on that change, does it?